src/lib.rs

```diff
@@ -6,70 +6,263 @@
 //! // Convert script
-//! let result_three: String = converter.traditional_to_simplified(traditional_text);
-//! println!("{}", result_three == simplified_text); // --> true
+//! let result_three = traditional_to_simplified(traditional_text);
+//! assert_eq!(result_three, simplified_text);
 //!
-//! let result_four: String = converter.simplified_to_traditional(simplified_text);
-//! println!("{}", result_four == traditional_text); // --> true
+//! let result_four = simplified_to_traditional(simplified_text);
+//! assert_eq!(result_four, traditional_text);
+//! ```
+#![cfg_attr(feature = "bench", feature(test))]
 
 extern crate bincode;
 
-mod character_converter;
-pub use self::character_converter::Converter as CharacterConverter;
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+use bincode::deserialize_from;
+use fst::raw::{Fst, Output};
+use once_cell::sync::Lazy;
+
+static T2S: Lazy<HashMap<String, String>> =
+    Lazy::new(|| deserialize_from(&include_bytes!("../data/t2s.profile")[..]).unwrap());
+static S2T: Lazy<HashMap<String, String>> =
+    Lazy::new(|| deserialize_from(&include_bytes!("../data/s2t.profile")[..]).unwrap());
+
+// create an fst containing all the keys
+static T2S_FST: Lazy<Fst<Vec<u8>>> = Lazy::new(|| {
+    let mut keys: Vec<_> = T2S.keys().collect();
+    keys.sort_unstable();
+    Fst::from_iter_set(keys).unwrap()
+});
+
+// create an fst containing all the keys
+static S2T_FST: Lazy<Fst<Vec<u8>>> = Lazy::new(|| {
+    let mut keys: Vec<_> = S2T.keys().collect();
+    keys.sort_unstable();
+    Fst::from_iter_set(keys).unwrap()
+});
+
+fn is_script(
+    raw: &str,
+    mapping: &HashMap<String, String>,
+    backup: &HashMap<String, String>,
+) -> bool {
+    for character in raw.chars() {
+        if mapping.get(&character.to_string()).is_none() {
+            if backup.get(&character.to_string()).is_some() {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+pub fn is_traditional(raw: &str) -> bool {
+    is_script(raw, &T2S, &S2T)
+}
+
+pub fn is_simplified(raw: &str) -> bool {
+    is_script(raw, &S2T, &T2S)
+}
+
+fn convert_script<'a>(
+    raw: &'a str,
+    mapping: &HashMap<String, String>,
+    fst: &Fst<Vec<u8>>,
+) -> Cow<'a, str> {
+    let mut converted_characters: Option<String> = None;
+    let mut skip_bytes = 0;
+
+    while skip_bytes < raw.len() {
+        let tailstr = &raw[skip_bytes..];
+
+        match find_longest_prefix(fst, tailstr.as_bytes()) {
+            Some((_, length)) => {
+                let tailstr = &tailstr[..length];
+                let mapped = mapping.get(tailstr).unwrap();
+                if let Some(converted_characters) = converted_characters.as_mut() {
+                    converted_characters.push_str(mapped);
+                } else if tailstr != mapped {
+                    converted_characters
+                        .get_or_insert_with(|| raw[..skip_bytes].to_string())
+                        .push_str(mapped);
```

Comment:
There is a case where the `raw[..skip_bytes].to_string()` string can create a string with not enough capacity and pushing the `mapped` one can force a reallocation, we can avoid this by being sure to reserve enough space.

```suggestion
                    let mut string = String::with_capacity(skip_bytes + mapped.len());
                    string.push_str(&raw[..skip_bytes]);
                    string.push_str(mapped);
                    converted_characters = Some(string);
```